### PR TITLE
Initial build system

### PR DESCRIPTION
**Pull request by jakearchibald**
Here's where I'm at with the build system. I'd like some early feedback.

It also feels like there's enough here to unblock us in terms of building. However, I can't get watch working, which feels essential, and I'm worried that it can't be done in the current version of Rollup. I've filed https://github.com/rollup/rollup/issues/2964 to see what's going on.

It could be that I'm pushing Rollup beyond what it's (currently) designed for. I'm going to take a look at Webpack and see if there's a smoother path there (although I'm worried about the issues we've had with webpack & workers in the past).

### CODE REVIEW

**Comment by surma** [comment]
I’ll take a look tomorrow! Thanks for going through this.

FWIW, personally I don’t find watch essential for this kind of work, but I might be the odd one out here. Could it be sufficient to set up a `nodemon` before this becomes a blocker or even — _gasp_ — a gateway drug for webpack?

**Comment by jakearchibald** [comment]
I found the lack of 'watch' on PROXX to be a bit of a productivity negative, as it took so long to rebuild. In my experience the benefit of watch scripts is in the incremental builds. I realise that you don't need watch for this, but it seems like the tools we use only do incremental builds in watch mode.

I'm similarly meh about webpack given our previous experience, but I also feel like I've had to bend Rollup a lot here. I'm going to have a play with it, just in case it turns out to be the better answer.

**Comment by jakearchibald** [comment]
I tried to find out how dynamically add entry points & emit assets in webpack and yea no let's not use webpack.

File: lib/html-entry-plugin.js [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        watcher.on('ready', () => {
+          if (!options.watch) watcher.close();
+          resolve();
+        });
+      });
+    },
+    generateBundle(_, bundle) {
+      // Remove empty.js
+      const emptyPath = resolve(__dirname, 'lib', 'empty.js');
+      for (const entry of Object.values(bundle)) {
```
Comment (surma) [inline lib/html-entry-plugin.js]: Is `entry.fileName` _always_ the key for `bundle`? I’d feel a bit better if we iterated of `Object.entries()` :D

Comment (jakearchibald) [reply lib/html-entry-plugin.js]: I've never seen them differ, but happy to be extra safe.

File: lib/html-entry-plugin.js [hunk 2]
```diff
@@ -0,0 +1,129 @@
+import fs from 'fs';
+import { promisify } from 'util';
```
Comment (robdodson) [inline lib/html-entry-plugin.js]: is this used?

Comment (jakearchibald) [reply lib/html-entry-plugin.js]: lol nope. Good spot. I'll remove it.

File: lib/html-entry-plugin.js [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+import { resolve, join } from 'path';
+
+import { minify as minifyHTML } from 'html-minifier';
+import chokidar from 'chokidar';
+
+function processHTML(rollup, htmlPath, assetCache) {
+  let html = fs.readFileSync(htmlPath, { encoding: 'utf8' });
```
[the remark this answers is not in the document]
Comment (jakearchibald) [reply lib/html-entry-plugin.js]: I've improved this in places, but not so much in the HTML entry plugin. I'll deal with this later: https://github.com/GoogleChrome/devsummit/issues/106

File: src/_includes/main/index.njk [hunk 4]
```diff
@@ -0,0 +1,16 @@
+{%- set css = "/_includes/main/style.css" -%}
+<!doctype html>
+<html lang="en">
+  <head>
+    <meta charset="utf-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
```
Comment (mathiasbynens) [inline src/_includes/main/index.njk]: I double-checked the math and 1.0 === 1

Comment (jakearchibald) [reply src/_includes/main/index.njk]: haha there was a time (maybe still) that this syntax was so fragile that I always copy/paste it. Will fix. Thanks for taking a look!

File: lib/html-entry-plugin.js [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+};
+
+export default function htmlEntryPlugin(userOptions = {}) {
+  let htmlOutput;
+  let assetCache;
+  const options = { ...defaultOptions, ...userOptions };
+
+  return {
+    name: 'html-entry-plugin',
+    async buildStart() {
```
Comment (surma) [inline lib/html-entry-plugin.js]: can this be moved to the variables initialization now?

File: rollup.config.js [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+  return {
+    output: {
+      dir: 'build',
+      format: 'esm',
+      assetFileNames: '[name]-[hash][extname]',
+    },
+    watch: { clearScreen: false },
+    plugins: [
+      globInputPlugin('.build-tmp/**/*.html'),
```
Comment (surma) [inline rollup.config.js]: Would it be cleaner if the `globInputPlugin` mapped over `options.input`? That way the inputs would remain in the `input` key in the config.

We can also kick this one down the road, not high prio.

File: lib/html-entry-plugin.js (hunk 5 above)
Comment (jakearchibald) [reply lib/html-entry-plugin.js]: Nah, you get many `buildStart`s in watch mode, but only one initialization.

File: rollup.config.js (hunk 6 above)
Comment (jakearchibald) [reply rollup.config.js]: I like the idea. I'll have a think about it. I'm not sure what to do with the object form of `input`.

**Pull request merged**